test/boost/alternator_unit_test.cc

```diff
@@ -86,3 +83,86 @@ BOOST_AUTO_TEST_CASE(test_allocator_fail_gracefully) {
+        int precision;
+    };
+    std::vector<expected> tests = {
+        // number     magnitude, precision
+        {"0",         0, 0},
+        {"0e10",      0, 0},
+        {"0e-10",     0, 0},
+        {"0e+10",     0, 0},
+        {"0.0",       0, 0},
+        {"0.00e10",   0, 0},
+        {"1",         0, 1},
+        {"12.",       1, 2},
+        {"1.1",       0, 2},
+        {"12.3",      1, 3},
+        {"12.300",    1, 3},
+        {"0.3",       -1, 1},
+        {".3",        -1, 1},
+        {"3e-1",      -1, 1},
+        {"0.00012",   -4, 2},
+        {"1.2e-4",    -4, 2},
+        {"1.2E-4",    -4, 2},
+        {"12.345e50", 51, 5},
+        {"12.345e-50",-49, 5},
+        {"123000000", 8, 3},
+        {"123000000.000e+5", 13, 3},
+        {"10.01",     1, 4},
+        {"1.001e1",   1, 4},
+        {"1e5",       5, 1},
+        {"1e+5",      5, 1},
+        {"1e-5",      -5, 1},
+        {"123e-7",    -5, 3},
+        // These are important edge cases: DynamoDB considers 1e126 to be
+        // overflowing but 9.9999e125 is considered to have magnitude 125
+        // and ok. Conversely, 1e-131 is underflowing and 0.9e-130 is too.
+        {"9.99999e125", 125, 6},
+        {"0.99999e-130", -131, 5},
+        {"0.9e-130", -131, 1},
+        // Although 1e1000 is not allowed, 0e0000 is allowed - it's just 0.
+        {"0e1000",    0, 0},
+    };
+    // prefixes that should do nothing to a number
+    std::vector<std::string> prefixes = {
+        "",
+        "0",
+        "+",
+        "-",
+        "+0000",
+        "-0000"
+    };
+    for (expected test : tests) {
+        for (std::string prefix : prefixes) {
+            std::string number = prefix + test.number;
+            auto res = alternator::internal::get_magnitude_and_precision(number);
+            BOOST_CHECK_MESSAGE(res.magnitude == test.magnitude,
+                format("{}: expected magnitude {}, got {}", number, test.magnitude, res.magnitude));
+            BOOST_CHECK_MESSAGE(res.precision == test.precision,
+                format("{}: expected precision {}, got {}", number, test.precision, res.precision));
+        }
+    }
+    // Huge exponents like 1e1000000 are not guaranteed to return that
+    // specific number as magnitude, but is guaranteed to return some
+    // other high magnitude that the caller can complain is excessive.
+    auto res = alternator::internal::get_magnitude_and_precision("1e1000000");
+    BOOST_CHECK(res.magnitude > 1000);
+    res = alternator::internal::get_magnitude_and_precision("1e-1000000");
+    BOOST_CHECK(res.magnitude < -1000);
+    // Even if an exponent so huge that it doesn't even fit in a 32-bit
+    // integer, we shouldn't fail to recognize its excessive magnitude:
+    res = alternator::internal::get_magnitude_and_precision("1e1000000000000");
+    BOOST_CHECK(res.magnitude > 1000);
+    res = alternator::internal::get_magnitude_and_precision("1e-1000000000000");
+    BOOST_CHECK(res.magnitude < -1000);
+}
```

Comment:
I think it would be worth adding some negative checks too, for invalid numeric strings, to ensure the validator will correctly recognize it.

Reply:
The function get_magnitude_and_precision() is written to assume that the number was already parsed correctly into a big_decimal() - the caller (Alternator) first creates a big_decimal() and only if it succeeds, validates the magnitude and precision).

As I explained in a comment, in a perfect world, the function would have used the already-parsed big_decimal and not reparse the string - but I didn't manage to do this  :-( Just now I realized that boost::multiprecision::cpp_int (which we use) implements msb() and lsb() functions - https://www.boost.org/doc/libs/1_55_0/libs/multiprecision/doc/html/boost_multiprecision/tut/gen_int.html - maybe I can rewrite my function in terms of those functions, working on a parsed big_decimal! I'll consider doing this, but I think maybe I already spent too much time on this issue already :-(

Reply:
Ok, queued. It can be improved it later if you find the time or reason.